cpp/arcae/shared_fcntl_mutex.h

```diff
@@ -0,0 +1,160 @@
+#ifndef ARCAE_SHARED_FCNTL_MUTEX_H
+#define ARCAE_SHARED_FCNTL_MUTEX_H
+
+#include <cerrno>
+
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <string_view>
+
+#if defined(_WIN32) || defined(_WIN64)
+#error Posix fcntl support required
+#else
+#include <fcntl.h>
+#include <unistd.h>
+#endif
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+#include <arrow/util/logging.h>
+
+using namespace std::literals;
+
+namespace arcae {
+
+using FcntlLockType = decltype(flock::l_type);
+
+// Base locking class
+class BaseSharedFcntlMutex {
+ public:
+  virtual arrow::Status lock() = 0;
+  virtual arrow::Status lock_shared() = 0;
+  virtual arrow::Result<bool> try_lock() = 0;
+  virtual arrow::Result<bool> try_lock_shared() = 0;
+
+  virtual void unlock() = 0;
+  virtual void unlock_shared() = 0;
+  virtual ~BaseSharedFcntlMutex() {};
+};
+
+// Noop class, mostly for use with in memory tables
+class NullSharedFcntlMutex : public BaseSharedFcntlMutex {
+ public:
+  virtual arrow::Status lock() override { return arrow::Status::OK(); }
+  virtual arrow::Status lock_shared() override { return arrow::Status::OK(); }
+
+  virtual arrow::Result<bool> try_lock() override { return true; }
+  virtual arrow::Result<bool> try_lock_shared() override { return true; }
+
+  virtual void unlock() override {}
+  virtual void unlock_shared() override {};
+};
+
+// Two part lock guarding access to a CASA table across multiple processes
+//
+// It's interface is somewhat similar to C++ standard library mutexes
+// but methods may return arrow::{Status,Result} instead.
+//
+// It combines:
+//
+// 1. A per-process lock that provides multiple-reader single-writer (MRSW)
+//    access to an underlying
+// 2. fcntl process lock that coordinates multiple-reader single-writer (MRSW)
+//    access across processes.
+//
+// (1) is implemented using a std::shared_mutex. By it's nature this restricts
+// access to the underlying fcntl lock to a single thread in the case of writes.
+// Further care must be taken to restrict access of multiple readers
+// to the underlying fcntl lock: This is accomplished with a std::mutex
+// and a reader counter.
+class SharedFcntlMutex : public BaseSharedFcntlMutex,
+                         std::enable_shared_from_this<SharedFcntlMutex> {
```

Comment:
Is there a particular reason you need `enable_shared_from_this`? I don't see you using `std::shared_from_this` anywhere.